File: hooks/scripts/cmd/post_edit.go
// PostToolUse(Edit/Write/MultiEdit): Track edits and remind to run code-simplifier.
package cmd

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"time"

	"github.com/spf13/cobra"
	"solon-hooks/internal/config"
	"solon-hooks/internal/git"
	"solon-hooks/internal/hookio"
)

var postEditCmd = &cobra.Command{
	Use:   "post-edit",
	Short: "Handle PostToolUse post-edit simplify reminder",
	RunE:  runPostEdit,
}

const (
	simplifySessionFile = "sl-simplify-session.json"
	editThreshold       = 5
	sessionTTLMs        = 2 * 60 * 60 * 1000  // 2 hours in ms
	reminderCooldownMs  = 10 * 60 * 1000       // 10 minutes in ms
)

type simplifySession struct {
	StartTime     int64    `json:"startTime"`
	EditCount     int      `json:"editCount"`
	ModifiedFiles []string `json:"modifiedFiles"`
	LastReminder  int64    `json:"lastReminder"`
	SimplifierRun bool     `json:"simplifierRun"`
}

func newSimplifySession() *simplifySession {
	return &simplifySession{
		StartTime:     time.Now().UnixMilli(),
		ModifiedFiles: []string{},
	}
}

func loadSimplifySession() *simplifySession {
	path := filepath.Join(os.TempDir(), simplifySessionFile)
	data, err := os.ReadFile(path)
	if err != nil {
		return newSimplifySession()
	}
	var s simplifySession
	if err := json.Unmarshal(data, &s); err != nil {
		return newSimplifySession()
	}
	// Reset if older than 2 hours
	if time.Now().UnixMilli()-s.StartTime > sessionTTLMs {
		return newSimplifySession()
	}
	return &s
}

func saveSimplifySession(s *simplifySession) {
	path := filepath.Join(os.TempDir(), simplifySessionFile)
	data, err := json.MarshalIndent(s, "", "  ")
	if err != nil {
		return
	}
	_ = os.WriteFile(path, data, 0644)
}

func runPostEdit(cmd *cobra.Command, args []string) error {
	if !config.IsHookEnabled("post-edit-simplify-reminder") {
		os.Exit(0)
	}

	var input hookio.PostToolUseInput
	if err := hookio.ReadInput(&input); err != nil {
		hookio.WriteOutput(map[string]interface{}{"continue": true})
		return nil
	}

	editTools := map[string]bool{"Edit": true, "Write": true, "MultiEdit": true}
	if !editTools[input.ToolName] {
		hookio.WriteOutput(map[string]interface{}{"continue": true})
		return nil
	}

	cwd := input.CWD
	if cwd == "" {
		cwd, _ = os.Getwd()
	}
	git.InvalidateCache(cwd)

	s := loadSimplifySession()
	s.EditCount++

	// Extract file path from tool input
	filePath := ""
	if fp, ok := input.ToolInput["file_path"].(string); ok && fp != "" {
		filePath = fp
	} else if p, ok := input.ToolInput["path"].(string); ok && p != "" {
		filePath = p
	}
	if filePath != "" {
		found := false
		for _, f := range s.ModifiedFiles {
			if f == filePath {
				found = true
				break
			}
		}
		if !found {
			s.ModifiedFiles = append(s.ModifiedFiles, filePath)
		}
	}

	now := time.Now().UnixMilli()
	shouldRemind := s.EditCount >= editThreshold &&
		!s.SimplifierRun &&
		now-s.LastReminder > reminderCooldownMs

	output := map[string]interface{}{"continue": true}
	if shouldRemind {
		s.LastReminder = now
		output["additionalContext"] = fmt.Sprintf(
			"\n\n[Code Simplification Reminder] You have modified %d files in this session. "+
				"Consider using the `code-simplifier` agent to refine recent changes before proceeding to code review. "+
				"This is a MANDATORY step in the workflow.",
			len(s.ModifiedFiles),
		)
	}

	saveSimplifySession(s)
	hookio.WriteOutput(output)
	return nil
}

File: crates/solon-core/src/hooks/post_edit.rs
use crate::hooks::{get_str, is_hook_enabled};
/// PostToolUse(Edit/Write/MultiEdit): track edits, remind to run code-simplifier after threshold.
use anyhow::Result;
use serde_json::Value;

const SIMPLIFY_SESSION_FILE: &str = "sl-simplify-session.json";
const EDIT_THRESHOLD: usize = 5;
const SESSION_TTL_MS: i64 = 2 * 60 * 60 * 1000; // 2 hours
const REMINDER_COOLDOWN_MS: i64 = 10 * 60 * 1000; // 10 minutes

#[derive(serde::Serialize, serde::Deserialize)]
struct SimplifySession {
    start_time: i64,
    edit_count: usize,
    modified_files: Vec<String>,
    last_reminder: i64,
    simplifier_run: bool,
}

impl SimplifySession {
    fn new() -> Self {
        Self {
            start_time: chrono::Utc::now().timestamp_millis(),
            edit_count: 0,
            modified_files: Vec::new(),
            last_reminder: 0,
            simplifier_run: false,
        }
    }
}

pub fn run() -> Result<()> {
    if !is_hook_enabled("post-edit-simplify-reminder") {
        crate::hooks::write_output(&serde_json::json!({"continue": true}));
        return Ok(());
    }

    let input = match crate::hooks::read_hook_input() {
        Ok(v) => v,
        Err(_) => {
            crate::hooks::write_output(&serde_json::json!({"continue": true}));
            return Ok(());
        }
    };

    let tool_name = get_str(&input, "tool_name");
    if !matches!(tool_name, "Edit" | "Write" | "MultiEdit") {
        crate::hooks::write_output(&serde_json::json!({"continue": true}));
        return Ok(());
    }

    // Invalidate git cache for the cwd
    let cwd = {
        let c = get_str(&input, "cwd").to_string();
        if c.is_empty() {
            std::env::current_dir()
                .unwrap_or_default()
                .to_string_lossy()
                .to_string()
        } else {
            c
        }
    };
    invalidate_git_cache(&cwd);

    let mut session = load_session();
    session.edit_count += 1;

    // Extract file path
    let file_path = input
        .get("tool_input")
        .and_then(|ti| ti.get("file_path").or_else(|| ti.get("path")))
        .and_then(|v| v.as_str())
        .filter(|s| !s.is_empty())
        .map(|s| s.to_string());

    if let Some(fp) = file_path {
        if !session.modified_files.contains(&fp) {
            session.modified_files.push(fp);
        }
    }

    let now = chrono::Utc::now().timestamp_millis();
    let should_remind = session.edit_count >= EDIT_THRESHOLD
        && !session.simplifier_run
        && now - session.last_reminder > REMINDER_COOLDOWN_MS;

    let mut output = serde_json::json!({"continue": true});
    if should_remind {
        session.last_reminder = now;
        output["additionalContext"] = Value::String(format!(
            "\n\n[Code Simplification Reminder] You have modified {} files in this session. \
             Consider using the `code-simplifier` agent to refine recent changes before proceeding to code review. \
             This is a MANDATORY step in the workflow.",
            session.modified_files.len()
        ));
    }

    save_session(&session);
    crate::hooks::write_output(&output);
    Ok(())
}

fn session_path() -> std::path::PathBuf {
    std::env::temp_dir().join(SIMPLIFY_SESSION_FILE)
}

fn load_session() -> SimplifySession {
    let data = match std::fs::read_to_string(session_path()) {
        Ok(d) => d,
        Err(_) => return SimplifySession::new(),
    };
    let s: SimplifySession = match serde_json::from_str(&data) {
        Ok(s) => s,
        Err(_) => return SimplifySession::new(),
    };
    let now = chrono::Utc::now().timestamp_millis();
    if now - s.start_time > SESSION_TTL_MS {
        return SimplifySession::new();
    }
    s
}

fn save_session(s: &SimplifySession) {
    if let Ok(data) = serde_json::to_string_pretty(s) {
        let _ = std::fs::write(session_path(), data);
    }
}

fn invalidate_git_cache(cwd: &str) {
    use std::collections::hash_map::DefaultHasher;
    use std::hash::{Hash, Hasher};
    let mut hasher = DefaultHasher::new();
    cwd.hash(&mut hasher);
    let hash = format!("{:x}", hasher.finish());
    let short = &hash[..8.min(hash.len())];
    let cache = std::env::temp_dir().join(format!("sl-git-cache-{}.json", short));
    let _ = std::fs::remove_file(cache);
}
